Sources/Bookmarks/BookmarkEntity.swift

```diff
@@ -143,6 +157,19 @@ public class BookmarkEntity: NSManagedObject {
         isFavorite = false
         favoriteFolder = nil
     }
+
+    public func markPendingDeletion() {
+        var queue: [BookmarkEntity] = [self]
+
+        while !queue.isEmpty {
+            let currentObject = queue.removeFirst()
+            currentObject.isPendingDeletion = true
+
+            if currentObject.isFolder {
+                queue.append(contentsOf: currentObject.childrenArray)
+            }
+        }
```

Comment:
Based on the privacy discussion, we should be cleaning up URL a title right? I wonder where is a good place to do that, but since we are already having business logic here (tree-handling) it may be as well be here?

Reply:
That’s right. Apologies that this code is half-baked. It doesn’t reflect the latest decisions like clearing URL and title. Let me fix it. Would you rather see tree handling in `BookmarkListViewModel.softDeleteBookmark`? The problem for me was that iOS and macOS use different APIs for bookmarks – iOS uses bookmark management code from BSK while macOS has its own BookmarkStore (with `remove(objectsWithUUIDs:)` method), so the common denominator is BookmarkEntity.

Still I believe that clearing URL and title is part of soft deletion, and it doesn’t make sense on its own without also setting `isPendingDeletion` flag – so it should be added here.

Reply:
Well, we had been clearing the folder contents with `cascade` delete rule before these changes, so let's keep it here, along with other soft-deletion logic.

The only problem I have, is that it binds three responsibilities here:
1. DB entity Deletion - this is expected.
2. Privacy constraints re stored objects (cleaning) - this is fine, part of the model conformation with high level sync mechanism.
3. Sync data validation constraints - this is something one may not see a direct connection with: we must ensure that the cleared out data still allows us to perform successful PATCH request, thus tying this with model validation on the backend. If endpoint were to expect non-optional title, that requirement would have to be respected here, which couples server validation with this place.

So my only worry is in the end #3 here, so I wonder if we should fall back to the model Spec in Asana and explicitly mention there that we tie any sync-model validation with validation that we have in the DB (for every field).

Reply:
We’ve agreed on a Zoom call that handling #3 is about updating docs and keeping our apps’ model validations the same as sync model validations. Hence I’ve removed the Core Data validation for non-nil bookmark URL.